crypto/provider_core.c

```diff
@@ -556,8 +556,29 @@ OSSL_PROVIDER *ossl_provider_new(OSSL_LIB_CTX *libctx, const char *name,
         template.init = init_function;
     }
 
+    if (params != NULL) {
+        int i;
+
+        template.parameters = sk_INFOPAIR_new_null();
+        if (template.parameters == NULL)
+            return NULL;
+
+        for (i = 0; params[i].key != NULL; i++) {
+            if (params[i].data_type != OSSL_PARAM_UTF8_STRING)
```

Comment:
So non-UTF* string params are not allowed?

This isn't really passing PARAMs to the provider, I was kind of expecting a OSSL_PROVIDER_set_params() call implemented on the provider side.  OSSL_PARAM is our define mechanism for communicating across the libcrypto/provider boundary.

Reply:
IMO that would be a bad mechanism. We want to be able to configure existing providers this way and they would not have this set_params() call. Why wouldn't we use the existing mechanism of providers calling back to libcrypto?

Reply:
On the other hand we do not necessarily have to convert params passed by the application in the load_ex() calls to INFOPAIRs. So we could support any types of OSSL_PARAMS here. We could drop the INFOPAIR type altogether and just create UTF8 params from the config file as well.

Reply:
Yes, limiting parameters to string type is intentional. 

I don't like the design here but I prefer to use the existing logic.

A separate problem is that when we explicitly pass the params to the application, we can't get them back.

Reply:
> Why wouldn't we use the existing mechanism of providers calling back to libcrypto?

We're trying to push parameters into providers.  How would a provider know _when_ to call back to libcrypto for changed settings?

Startup/loading isn't the only time we might want to change things.

Reply:
I think we don't plan to support change of the parameters just now

Reply:
> We're trying to push parameters into providers. How would a provider know _when_ to call back to libcrypto for changed settings?
> 
> Startup/loading isn't the only time we might want to change things.

Actually quite the opposite - we IMO do not want to change things after the provider initialization as that complicates provider implementation.

Or yeah, we might want to add the push-parameters-to-provider method eventually as well but that does not replace this proposal.

Reply:
I don't really see a reason for a push model.  The provider code can retrieve whatever parameter it desires at any time it needs it, I don't see how that's a bad thing...